File: packages/sdk-ts/src/page.ts
import type {
  LoadState,
  PageCDPEvent,
  PageEventName,
  PageClickParams,
  PageDragAndDropParams,
  PageKeyPressParams,
  PageNavigationOptions,
  PageRef,
  PageReloadParams,
  PageScreenshotOptions,
  PageSetExtraHTTPHeadersParams,
  PageSetViewportSizeParams,
  PageSnapshotOptions,
  SnapshotResult,
  PageTypeParams,
  PageWaitForSelectorParams,
  PageWaitForTimeoutParams,
} from "@browserbasehq/stagehand-protocol/types";
import {
  StagehandMethods,
  StagehandNotifications,
} from "@browserbasehq/stagehand-protocol/schema-registry";
import { decodeBase64 } from "./base64.js";
import { Locator } from "./locator.js";
import {
  type InitScriptSource,
  normalizeEvaluationExpression,
  normalizeInitScriptSource,
} from "./pageScripts.js";
import type { StagehandCommandClient } from "./commandClient.js";
import { Response } from "./response.js";
import { WebMCPTool } from "./webmcp.js";
import type { WebMCPToolsOptions } from "./clientSchemas.js";

export type ScreenshotOptions = Omit<PageScreenshotOptions, "mask"> & {
  mask?: Locator[];
  path?: string;
};

export type PageClickOptions = NonNullable<PageClickParams["options"]>;
export type PageDragAndDropOptions = NonNullable<PageDragAndDropParams["options"]>;
export type PageKeyPressOptions = NonNullable<PageKeyPressParams["options"]>;
export type PageReloadOptions = NonNullable<PageReloadParams["options"]>;
export type PageSetViewportSizeOptions = NonNullable<PageSetViewportSizeParams["options"]>;
export type PageTypeOptions = NonNullable<PageTypeParams["options"]>;
export type PageWaitForSelectorOptions = NonNullable<PageWaitForSelectorParams["options"]>;

export interface PageEventListener {
  (event: PageCDPEvent): unknown;
}

export class CDPSubscription {
  private unsubscribePromise: Promise<void> | undefined;

  constructor(
    private readonly rpcClient: StagehandCommandClient,
    private readonly subscriptionId: string,
    private readonly removeLocalListener: () => void,
    private readonly onDisposed: () => void,
  ) {}

  unsubscribe(): Promise<void> {
    this.unsubscribePromise ??= this.unsubscribeOnce().catch((error: unknown) => {
      this.unsubscribePromise = undefined;
      throw error;
    });
    return this.unsubscribePromise;
  }

  private async unsubscribeOnce(): Promise<void> {
    this.removeLocalListener();
    this.onDisposed();
    await this.rpcClient.send(StagehandMethods.pageOff, {
      subscriptionId: this.subscriptionId,
    });
  }
}

export class Page {
  currentRef: PageRef;
  private readonly eventSubscriptions = new Set<CDPSubscription>();

  constructor(
    readonly rpcClient: StagehandCommandClient,
    ref: PageRef,
  ) {
    this.currentRef = ref;
  }

  get pageId(): string {
    return this.currentRef.pageId;
  }

  get ref(): PageRef {
    return this.currentRef;
  }

  async goto(url: string, options?: PageNavigationOptions): Promise<Response | null> {
    const result = await this.rpcClient.send(StagehandMethods.pageGoto, {
      pageId: this.pageId,
      url,
      ...(options ? { options } : {}),
    });
    this.currentRef = result.page;
    return result.response === null ? null : new Response(this.rpcClient, result.response);
  }

  async reload(options?: PageReloadOptions): Promise<Response | null> {
    const result = await this.rpcClient.send(StagehandMethods.pageReload, {
      pageId: this.pageId,
      ...(options ? { options } : {}),
    });
    this.currentRef = result.page;
    return result.response === null ? null : new Response(this.rpcClient, result.response);
  }

  async goBack(options?: PageNavigationOptions): Promise<Response | null> {
    const result = await this.rpcClient.send(StagehandMethods.pageGoBack, {
      pageId: this.pageId,
      ...(options ? { options } : {}),
    });
    this.currentRef = result.page;
    return result.response === null ? null : new Response(this.rpcClient, result.response);
  }

  async goForward(options?: PageNavigationOptions): Promise<Response | null> {
    const result = await this.rpcClient.send(StagehandMethods.pageGoForward, {
      pageId: this.pageId,
      ...(options ? { options } : {}),
    });
    this.currentRef = result.page;
    return result.response === null ? null : new Response(this.rpcClient, result.response);
  }

  async click(x: number, y: number, options?: PageClickOptions): Promise<void> {
    await this.rpcClient.send(StagehandMethods.pageClick, {
      pageId: this.pageId,
      x,
      y,
      ...(options ? { options } : {}),
    });
  }

  async hover(x: number, y: number): Promise<void> {
    await this.rpcClient.send(StagehandMethods.pageHover, {
      pageId: this.pageId,
      x,
      y,
    });
  }

  async scroll(x: number, y: number, deltaX: number, deltaY: number): Promise<void> {
    await this.rpcClient.send(StagehandMethods.pageScroll, {
      pageId: this.pageId,
      x,
      y,
      deltaX,
      deltaY,
    });
  }

  async dragAndDrop(
    fromX: number,
    fromY: number,
    toX: number,
    toY: number,
    options?: PageDragAndDropOptions,
  ): Promise<void> {
    await this.rpcClient.send(StagehandMethods.pageDragAndDrop, {
      pageId: this.pageId,
      fromX,
      fromY,
      toX,
      toY,
      ...(options ? { options } : {}),
    });
  }

  async type(text: string, options?: PageTypeOptions): Promise<void> {
    await this.rpcClient.send(StagehandMethods.pageType, {
      pageId: this.pageId,
      text,
      ...(options ? { options } : {}),
    });
  }

  async keyPress(key: string, options?: PageKeyPressOptions): Promise<void> {
    await this.rpcClient.send(StagehandMethods.pageKeyPress, {
      pageId: this.pageId,
      key,
      ...(options ? { options } : {}),
    });
  }

  async evaluate<R = unknown, Arg = unknown>(
    expression: string | ((arg: Arg) => R | Promise<R>),
    arg?: Arg,
  ): Promise<R> {
    const result = await this.rpcClient.send(StagehandMethods.pageEvaluate, {
      pageId: this.pageId,
      expression: normalizeEvaluationExpression(expression, arg),
    });
    return result.value as R;
  }

  async addInitScript<Arg = unknown>(script: InitScriptSource<Arg>, arg?: Arg): Promise<void> {
    const source = await normalizeInitScriptSource(script, arg);
    await this.rpcClient.send(StagehandMethods.pageAddInitScript, {
      pageId: this.pageId,
      source,
    });
  }

  async on(event: PageEventName, listener: PageEventListener): Promise<CDPSubscription> {
    const subscriptionId = crypto.randomUUID();
    const removeNotificationListener = this.rpcClient.onNotification((notification) => {
      if (
        notification.method !== StagehandNotifications.pageCDPEvent.name ||
        notification.params.subscriptionId !== subscriptionId
      ) {
        return;
      }
      try {
        const result = listener(notification.params.event);
        if (result && typeof result === "object" && "then" in result) {
          void Promise.resolve(result).catch(reportPageEventListenerError);
        }
      } catch (error) {
        reportPageEventListenerError(error);
      }
    });

    let subscription: CDPSubscription;
    subscription = new CDPSubscription(
      this.rpcClient,
      subscriptionId,
      removeNotificationListener,
      () => this.eventSubscriptions.delete(subscription),
    );
    this.eventSubscriptions.add(subscription);

    try {
      await this.rpcClient.send(StagehandMethods.pageOn, {
        pageId: this.pageId,
        subscriptionId,
        event,
      });
      return subscription;
    } catch (error) {
      removeNotificationListener();
      this.eventSubscriptions.delete(subscription);
      throw error;
    }
  }

  async setExtraHTTPHeaders(headers: PageSetExtraHTTPHeadersParams["headers"]): Promise<void> {
    await this.rpcClient.send(StagehandMethods.pageSetExtraHTTPHeaders, {
      pageId: this.pageId,
      headers,
    });
  }

  async setViewportSize(
    width: number,
    height: number,
    options?: PageSetViewportSizeOptions,
  ): Promise<void> {
    await this.rpcClient.send(StagehandMethods.pageSetViewportSize, {
      pageId: this.pageId,
      width,
      height,
      ...(options ? { options } : {}),
    });
  }

  async waitForLoadState(state: LoadState, timeout?: number): Promise<void> {
    await this.rpcClient.send(StagehandMethods.pageWaitForLoadState, {
      pageId: this.pageId,
      state,
      ...(timeout === undefined ? {} : { timeout }),
    });
  }

  async waitForTimeout(ms: PageWaitForTimeoutParams["ms"]): Promise<void> {
    await this.rpcClient.send(StagehandMethods.pageWaitForTimeout, {
      pageId: this.pageId,
      ms,
    });
  }

  async waitForSelector(selector: string, options?: PageWaitForSelectorOptions): Promise<boolean> {
    const result = await this.rpcClient.send(StagehandMethods.pageWaitForSelector, {
      pageId: this.pageId,
      selector,
      ...(options ? { options } : {}),
    });
    return result.matched;
  }

  async screenshot(options?: ScreenshotOptions): Promise<Uint8Array> {
    const { path, mask, ...screenshotOptions } = options ?? {};
    const result = await this.rpcClient.send(StagehandMethods.pageScreenshot, {
      pageId: this.pageId,
      options: {
        ...screenshotOptions,
        ...(mask ? { mask: mask.map((locator) => locator.descriptor) } : {}),
      },
    });
    const bytes = decodeBase64(result.data, "page.screenshot");
    if (path) {
      const moduleName = "node:" + "fs/promises";
      const { writeFile } = (await import(/* @vite-ignore */ moduleName).catch(() => {
        throw new TypeError(
          "page.screenshot(): path is only supported in Node.js; omit path to receive screenshot bytes",
        );
      })) as typeof import("node:fs/promises");
      await writeFile(path, bytes);
    }
    return bytes;
  }

  async snapshot(options?: PageSnapshotOptions): Promise<SnapshotResult> {
    return await this.rpcClient.send(StagehandMethods.pageSnapshot, {
      pageId: this.pageId,
      ...(options ? { options } : {}),
    });
  }

  async tools(options?: WebMCPToolsOptions): Promise<WebMCPTool[]> {
    const result = await this.rpcClient.send(StagehandMethods.pageWebMCPTools, {
      pageId: this.pageId,
      ...(options ? { options } : {}),
    });
    return result.tools.map(
      (descriptor) => new WebMCPTool(this.rpcClient, this.pageId, descriptor),
    );
  }

  async url(): Promise<string> {
    return await this.rpcClient.send(StagehandMethods.pageUrl, {
      pageId: this.pageId,
    });
  }

  async title(): Promise<string> {
    return await this.rpcClient.send(StagehandMethods.pageTitle, {
      pageId: this.pageId,
    });
  }

  async close(): Promise<void> {
    await Promise.allSettled(
      [...this.eventSubscriptions].map((subscription) => subscription.unsubscribe()),
    );
    await this.rpcClient.send(StagehandMethods.pageClose, { pageId: this.pageId });
  }

  locator(selector: string): Locator {
    return new Locator(this.rpcClient, {
      pageId: this.pageId,
      selector,
    });
  }
}

function reportPageEventListenerError(error: unknown): void {
  process.emitWarning(error instanceof Error ? error.message : String(error), {
    code: "STAGEHAND_PAGE_EVENT_LISTENER_ERROR",
  });
}

File: packages/sdk-python/src/stagehand/page.py
from __future__ import annotations

import asyncio
import base64
import builtins
import inspect
from collections.abc import Awaitable, Callable, Mapping, Sequence
from pathlib import Path
from typing import Literal, TypeVar, cast, overload
from uuid import uuid4

from pydantic import JsonValue, TypeAdapter

from ._generated.input_types import PageEventName, PageScreenshotClip
from ._generated.models import (
    Animations,
    Caret,
    LoadState,
    MouseButton,
    PageAddInitScriptParams,
    PageCDPEvent,
    PageCDPEventNotification,
    PageClickOptions,
    PageClickParams,
    PageCloseResult,
    PageDragAndDropOptions,
    PageDragAndDropParams,
    PageEvaluateParams,
    PageEvaluateResult,
    PageGoBackParams,
    PageGoForwardParams,
    PageGotoParams,
    PageHoverParams,
    PageIdParams,
    PageKeyPressOptions,
    PageKeyPressParams,
    PageNavigationOptions,
    PageNavigationResult,
    PageOffParams,
    PageOnParams,
    PageRef,
    PageReloadOptions,
    PageReloadParams,
    PageScreenshotOptions,
    PageScreenshotParams,
    PageScreenshotResult,
    PageScrollParams,
    PageSetExtraHTTPHeadersParams,
    PageSetViewportSizeOptions,
    PageSetViewportSizeParams,
    PageSnapshotOptions,
    PageSnapshotParams,
    PageTitleResult,
    PageTypeOptions,
    PageTypeParams,
    PageUrlResult,
    PageVoidResult,
    PageWaitForLoadStateParams,
    PageWaitForSelectorOptions,
    PageWaitForSelectorParams,
    PageWaitForSelectorResult,
    PageWaitForTimeoutParams,
    PageWebMCPToolsParams,
    PageWebMCPToolsResult,
    Scale,
    SnapshotResult,
    State,
    WebMCPToolsOptions,
)
from ._generated.models import (
    Type as ScreenshotType,
)
from .locator import Locator
from .response import Response
from .rpc_client import RPCClient
from .webmcp import WebMCPTool

EvaluateResult = TypeVar("EvaluateResult")
PageEventListener = Callable[[PageCDPEvent], object | Awaitable[object]]


class CDPSubscription:
    def __init__(
        self,
        rpc_client: RPCClient,
        subscription_id: str,
        remove_local_listener: Callable[[], None],
        on_disposed: Callable[[], None],
    ) -> None:
        self._rpc_client = rpc_client
        self._subscription_id = subscription_id
        self._remove_local_listener = remove_local_listener
        self._on_disposed = on_disposed
        self._unsubscribe_task: asyncio.Task[None] | None = None

    async def unsubscribe(self) -> None:
        if self._unsubscribe_task is None:

            async def run() -> None:
                await self._unsubscribe_once()

            self._unsubscribe_task = asyncio.create_task(run())
        await self._unsubscribe_task

    async def _unsubscribe_once(self) -> None:
        self._remove_local_listener()
        self._on_disposed()
        await self._rpc_client.send(
            "page.off",
            PageOffParams(subscription_id=self._subscription_id),
            PageVoidResult,
        )


class Page:
    def __init__(self, rpc_client: RPCClient, ref: PageRef) -> None:
        self._rpc_client = rpc_client
        self._ref = ref
        self._event_subscriptions: set[CDPSubscription] = set()

    @property
    def page_id(self) -> str:
        return self._ref.page_id

    @property
    def ref(self) -> PageRef:
        return self._ref

    async def goto(
        self,
        url: str,
        *,
        wait_until: LoadState | Literal["load", "domcontentloaded", "networkidle"] | None = None,
        timeout: int | None = None,
    ) -> Response | None:
        options = {
            name: value
            for name, value in (("wait_until", wait_until), ("timeout", timeout))
            if value is not None
        }
        params = PageGotoParams.model_validate({
            "page_id": self.page_id,
            "url": url,
            **({"options": options} if options else {}),
        })
        result = await self._rpc_client.send("page.goto", params, PageNavigationResult)
        self._ref = result.page
        return Response(self._rpc_client, result.response) if result.response is not None else None

    async def reload(
        self,
        *,
        wait_until: LoadState | Literal["load", "domcontentloaded", "networkidle"] | None = None,
        timeout: int | None = None,
        ignore_cache: bool | None = None,
    ) -> Response | None:
        params = PageReloadParams(page_id=self.page_id)
        options = PageReloadOptions.model_validate({
            name: value
            for name, value in (
                ("wait_until", wait_until),
                ("timeout", timeout),
                ("ignore_cache", ignore_cache),
            )
            if value is not None
        })
        if options.model_fields_set:
            params.options = options
        result = await self._rpc_client.send("page.reload", params, PageNavigationResult)
        self._ref = result.page
        return Response(self._rpc_client, result.response) if result.response is not None else None

    async def go_back(
        self,
        *,
        wait_until: LoadState | Literal["load", "domcontentloaded", "networkidle"] | None = None,
        timeout: int | None = None,
    ) -> Response | None:
        params = PageGoBackParams(page_id=self.page_id)
        options = PageNavigationOptions.model_validate({
            name: value
            for name, value in (("wait_until", wait_until), ("timeout", timeout))
            if value is not None
        })
        if options.model_fields_set:
            params.options = options
        result = await self._rpc_client.send("page.go_back", params, PageNavigationResult)
        self._ref = result.page
        return Response(self._rpc_client, result.response) if result.response is not None else None

    async def go_forward(
        self,
        *,
        wait_until: LoadState | Literal["load", "domcontentloaded", "networkidle"] | None = None,
        timeout: int | None = None,
    ) -> Response | None:
        params = PageGoForwardParams(page_id=self.page_id)
        options = PageNavigationOptions.model_validate({
            name: value
            for name, value in (("wait_until", wait_until), ("timeout", timeout))
            if value is not None
        })
        if options.model_fields_set:
            params.options = options
        result = await self._rpc_client.send("page.go_forward", params, PageNavigationResult)
        self._ref = result.page
        return Response(self._rpc_client, result.response) if result.response is not None else None

    async def click(
        self,
        x: float,
        y: float,
        *,
        button: MouseButton | Literal["left", "right", "middle"] | None = None,
        click_count: int | None = None,
    ) -> None:
        params = PageClickParams(page_id=self.page_id, x=x, y=y)
        options = PageClickOptions.model_validate({
            name: value
            for name, value in (
                ("button", button),
                ("click_count", click_count),
            )
            if value is not None
        })
        if options.model_fields_set:
            params.options = options
        await self._rpc_client.send("page.click", params, PageVoidResult)

    async def hover(
        self,
        x: float,
        y: float,
    ) -> None:
        params = PageHoverParams(page_id=self.page_id, x=x, y=y)
        await self._rpc_client.send("page.hover", params, PageVoidResult)

    async def scroll(
        self,
        x: float,
        y: float,
        delta_x: float,
        delta_y: float,
    ) -> None:
        params = PageScrollParams(
            page_id=self.page_id,
            x=x,
            y=y,
            delta_x=delta_x,
            delta_y=delta_y,
        )
        await self._rpc_client.send("page.scroll", params, PageVoidResult)

    async def drag_and_drop(
        self,
        from_x: float,
        from_y: float,
        to_x: float,
        to_y: float,
        *,
        button: MouseButton | Literal["left", "right", "middle"] | None = None,
        steps: int | None = None,
        delay: float | None = None,
    ) -> None:
        params = PageDragAndDropParams(
            page_id=self.page_id,
            from_x=from_x,
            from_y=from_y,
            to_x=to_x,
            to_y=to_y,
        )
        options = PageDragAndDropOptions.model_validate({
            name: value
            for name, value in (
                ("button", button),
                ("steps", steps),
                ("delay", delay),
            )
            if value is not None
        })
        if options.model_fields_set:
            params.options = options
        await self._rpc_client.send(
            "page.drag_and_drop",
            params,
            PageVoidResult,
        )

    async def type(
        self,
        text: str,
        *,
        delay: float | None = None,
        with_mistakes: bool | None = None,
    ) -> None:
        params = PageTypeParams(page_id=self.page_id, text=text)
        options = PageTypeOptions.model_validate({
            name: value
            for name, value in (("delay", delay), ("with_mistakes", with_mistakes))
            if value is not None
        })
        if options.model_fields_set:
            params.options = options
        await self._rpc_client.send("page.type", params, PageVoidResult)

    async def key_press(self, key: str, *, delay: float | None = None) -> None:
        params = PageKeyPressParams(page_id=self.page_id, key=key)
        if delay is not None:
            params.options = PageKeyPressOptions(delay=delay)
        await self._rpc_client.send("page.key_press", params, PageVoidResult)

    @overload
    async def evaluate(self, expression: str) -> JsonValue: ...

    @overload
    async def evaluate(
        self,
        expression: str,
        *,
        result_type: builtins.type[EvaluateResult],
    ) -> EvaluateResult: ...

    async def evaluate(
        self,
        expression: str,
        *,
        result_type: builtins.type[EvaluateResult] | None = None,
    ) -> JsonValue | EvaluateResult:
        result = await self._rpc_client.send(
            "page.evaluate",
            PageEvaluateParams(page_id=self.page_id, expression=expression),
            PageEvaluateResult,
        )
        value = None if result.value is None else result.value.model_dump()
        if result_type is None:
            return cast(JsonValue, value)
        return TypeAdapter(result_type).validate_python(value, strict=True)

    async def add_init_script(self, source: str | Path) -> None:
        if isinstance(source, Path):
            source_url = str(source).replace("\n", "")
            script = f"{source.read_text()}\n//# sourceURL={source_url}"
        else:
            script = source
        await self._rpc_client.send(
            "page.add_init_script",
            PageAddInitScriptParams(page_id=self.page_id, source=script),
            PageVoidResult,
        )

    async def on(
        self,
        event: PageEventName,
        listener: PageEventListener,
    ) -> CDPSubscription:
        subscription_id = uuid4().hex

        async def notify(notification: PageCDPEventNotification) -> None:
            if notification.subscription_id != subscription_id:
                return
            try:
                result = listener(notification.event)
                if inspect.isawaitable(result):
                    await result
            except Exception as error:
                asyncio.get_running_loop().call_exception_handler({
                    "message": "Stagehand page event listener failed",
                    "exception": error,
                })

        remove_notification_listener = self._rpc_client.on_notification(
            "page.cdp_event",
            PageCDPEventNotification,
            notify,
        )

        subscription: CDPSubscription
        subscription = CDPSubscription(
            self._rpc_client,
            subscription_id,
            remove_notification_listener,
            lambda: self._event_subscriptions.discard(subscription),
        )
        self._event_subscriptions.add(subscription)
        try:
            await self._rpc_client.send(
                "page.on",
                PageOnParams.model_validate({
                    "page_id": self.page_id,
                    "subscription_id": subscription_id,
                    "event": event,
                }),
                PageVoidResult,
            )
        except Exception:
            remove_notification_listener()
            self._event_subscriptions.discard(subscription)
            raise
        return subscription

    async def set_extra_http_headers(self, headers: Mapping[str, str]) -> None:
        await self._rpc_client.send(
            "page.set_extra_http_headers",
            PageSetExtraHTTPHeadersParams(page_id=self.page_id, headers=dict(headers)),
            PageVoidResult,
        )

    async def set_viewport_size(
        self,
        width: int,
        height: int,
        *,
        device_scale_factor: float | None = None,
    ) -> None:
        params = PageSetViewportSizeParams(page_id=self.page_id, width=width, height=height)
        if device_scale_factor is not None:
            params.options = PageSetViewportSizeOptions(
                device_scale_factor=device_scale_factor,
            )
        await self._rpc_client.send("page.set_viewport_size", params, PageVoidResult)

    async def wait_for_load_state(
        self,
        state: LoadState | Literal["load", "domcontentloaded", "networkidle"],
        *,
        timeout: int | None = None,
    ) -> None:
        params = PageWaitForLoadStateParams.model_validate({
            "page_id": self.page_id,
            "state": state,
        })
        if timeout is not None:
            params.timeout = timeout
        await self._rpc_client.send("page.wait_for_load_state", params, PageVoidResult)

    async def wait_for_timeout(self, ms: int) -> None:
        await self._rpc_client.send(
            "page.wait_for_timeout",
            PageWaitForTimeoutParams(page_id=self.page_id, ms=ms),
            PageVoidResult,
        )

    async def wait_for_selector(
        self,
        selector: str,
        *,
        state: State | Literal["attached", "detached", "visible", "hidden"] | None = None,
        timeout: int | None = None,
        pierce_shadow: bool | None = None,
    ) -> bool:
        params = PageWaitForSelectorParams(page_id=self.page_id, selector=selector)
        options = PageWaitForSelectorOptions.model_validate({
            name: value
            for name, value in (
                ("state", state),
                ("timeout", timeout),
                ("pierce_shadow", pierce_shadow),
            )
            if value is not None
        })
        if options.model_fields_set:
            params.options = options
        result = await self._rpc_client.send(
            "page.wait_for_selector",
            params,
            PageWaitForSelectorResult,
        )
        return result.matched

    async def screenshot(
        self,
        *,
        animations: Animations | Literal["disabled", "allow"] | None = None,
        caret: Caret | Literal["hide", "initial"] | None = None,
        clip: PageScreenshotClip | None = None,
        full_page: bool | None = None,
        path: str | Path | None = None,
        mask: Sequence[Locator] | None = None,
        mask_color: str | None = None,
        omit_background: bool | None = None,
        quality: int | None = None,
        scale: Scale | Literal["css", "device"] | None = None,
        style: str | None = None,
        timeout: float | None = None,
        type: ScreenshotType | Literal["png", "jpeg"] | None = None,
    ) -> bytes:
        params = PageScreenshotParams(page_id=self.page_id)
        options = PageScreenshotOptions.model_validate({
            name: value
            for name, value in (
                ("animations", animations),
                ("caret", caret),
                ("clip", clip),
                ("full_page", full_page),
                (
                    "mask",
                    [locator.descriptor for locator in mask] if mask is not None else None,
                ),
                ("mask_color", mask_color),
                ("omit_background", omit_background),
                ("quality", quality),
                ("scale", scale),
                ("style", style),
                ("timeout", timeout),
                ("type", type),
            )
            if value is not None
        })
        if options.model_fields_set:
            params.options = options
        result = await self._rpc_client.send(
            "page.screenshot",
            params,
            PageScreenshotResult,
        )
        data = base64.b64decode(result.data, validate=True)
        if path is not None:
            Path(path).write_bytes(data)
        return data

    async def snapshot(self, *, include_iframes: bool | None = None) -> SnapshotResult:
        params = PageSnapshotParams(page_id=self.page_id)
        if include_iframes is not None:
            params.options = PageSnapshotOptions(include_iframes=include_iframes)
        return await self._rpc_client.send("page.snapshot", params, SnapshotResult)

    async def tools(self, *, timeout: float | None = None) -> list[WebMCPTool]:
        params = PageWebMCPToolsParams(page_id=self.page_id)
        if timeout is not None:
            params.options = WebMCPToolsOptions(timeout=timeout)
        result = await self._rpc_client.send(
            "page.webmcp_tools",
            params,
            PageWebMCPToolsResult,
        )
        return [WebMCPTool(self._rpc_client, self.page_id, tool) for tool in result.tools]

    async def url(self) -> str:
        return await self._rpc_client.send(
            "page.url",
            PageIdParams(page_id=self.page_id),
            PageUrlResult,
        )

    async def title(self) -> str:
        return await self._rpc_client.send(
            "page.title",
            PageIdParams(page_id=self.page_id),
            PageTitleResult,
        )

    async def close(self) -> None:
        if self._event_subscriptions:
            await asyncio.gather(
                *(subscription.unsubscribe() for subscription in self._event_subscriptions),
                return_exceptions=True,
            )
        await self._rpc_client.send(
            "page.close",
            PageIdParams(page_id=self.page_id),
            PageCloseResult,
        )

    def locator(self, selector: str) -> Locator:
        return Locator(self._rpc_client, page_id=self.page_id, selector=selector)
